File: ruby/snakesladders/lib/snakesladders/services/game_service.rb
module Snakesladders
  module Services
    class GameService
      attr_accessor :board, :dice, :players, :snakes, :ladders

      def initialize(players, snakes, ladders, board_size, dice_range)
        self.players = players
        self.snakes = snakes
        self.ladders = ladders
        self.board = ::Snakesladders::Models::Board.new(board_size, snakes, ladders)
        self.dice = ::Snakesladders::Models::Dice.new(dice_range)
      end

      def run
        loop do
          player = players.shift
          take_turn(player, dice.roll)

          if player.current_cell == board.last_cell
            player.won = true
            puts "Player #{player.name} won"
            break
          else
            players << player
          end
        end
      end

      def take_turn(player, dice_value)
        if player.current_cell == -1
          player.current_cell = 1 if dice_value == 1
        elsif player.current_cell + dice_value > board.last_cell
          puts "[#{player.name}] Dice roll nullified as #{player.current_cell} + #{dice_value} > #{board.last_cell}"
        else
          player.current_cell += dice_value
          if board.snake_points[player.current_cell]
            destination = board.snake_points[player.current_cell]
            puts "[#{player.name}] cut by snake. Going back to #{destination}"
            player.current_cell = destination
          elsif board.ladder_points[player.current_cell]
            destination = board.ladder_points[player.current_cell]
            puts "[#{player.name}] promoted by ladder. Moving to #{destination}"
            player.current_cell = destination
          else
            puts "[#{player.name}] moving to #{player.current_cell} after dice value #{dice_value}"
          end
        end
      end
    end
  end
end

File: java/snakes-and-ladders/src/main/java/gagan93jtg/lld/services/GameService.java
package gagan93jtg.lld.services;

import gagan93jtg.lld.models.*;

import java.util.ArrayList;

public class GameService {
  private Board board;
  private Dice dice;
  private ArrayList<Player> players;
  private ArrayList<Snake> snakes;
  private ArrayList<Ladder> ladders;

  public GameService(ArrayList<Player> players, ArrayList<Snake> snakes, ArrayList<Ladder> ladders, int boardSize, int diceRange) {
    this.players = players;
    this.snakes = snakes;
    this.ladders = ladders;
    this.board = new Board(boardSize, snakes, ladders);
    this.dice = new Dice(diceRange);
  }

  public Board getBoard() {
    return board;
  }

  public void run() {
    while (true) {
      Player player = players.remove(0);
      takeTurn(player, dice.roll());

      if (player.getCurrentCell() == board.getLastCell()) {
        player.setWon(true);
        System.out.println("Player " + player.getName() + " is winner");
        break;
      } else {
        players.add(player);
      }
    }
  }

  public void takeTurn(Player player, int diceValue) {
    if (player.getCurrentCell() == -1) {
      if (diceValue == 1) {
        player.setCurrentCell(1);
      }
    } else if (player.getCurrentCell() + diceValue > board.getLastCell()) {
      System.out.println("Player current position " + player.getCurrentCell() + " plus dice value " + diceValue + " shall cross the board limit. Hence nullifying");
    } else {
      player.setCurrentCell(player.getCurrentCell() + diceValue);
      if (board.getSnakePoints().containsKey(player.getCurrentCell())) {
        int destination = board.getSnakePoints().get(player.getCurrentCell());
        System.out.println("[" + player.getName() + "] cut by snake. Now goes from " + player.getCurrentCell() + " to " + destination);
        player.setCurrentCell(destination);
      } else if (board.getLadderPoints().containsKey(player.getCurrentCell())) {
        int destination = board.getLadderPoints().get(player.getCurrentCell());
        System.out.println("[" + player.getName() + "] promoted by ladder. Now goes from " + player.getCurrentCell() + " to " + destination);
        player.setCurrentCell(destination);
      } else {
        System.out.println("[" + player.getName() + "] Current position " + player.getCurrentCell() + " after dice value : " + diceValue);
      }
    }
  }
}
